lib/resque/scheduler.rb

```diff
@@ -48,38 +56,46 @@ def run
         $stdout.sync = true
         $stderr.sync = true
 
-        was_master = nil
+        @was_master = nil
 
         begin
           @th = Thread.current
 
           # Now start the scheduling part of the loop.
-          loop do
-            begin
-              # Check on changes to master/child
-              @am_master = master?
-              if am_master != was_master
-                procline am_master ? 'Master scheduler' : 'Child scheduler'
-
-                # Load schedule because changed
-                reload_schedule!
-              end
+          loop { main_loop_once }
 
-              if am_master
-                handle_delayed_items
-                update_schedule if dynamic
-              end
-              was_master = am_master
-            rescue *INTERMITTENT_ERRORS => e
-              log! e.message
-              release_master_lock
+        rescue Interrupt
+          log 'Exiting'
+        end
+      end
+
+      def main_loop_once
+        begin
+          # Check on changes to master/child
+          @am_master = master?
+          if am_master != @was_master
+            procline am_master ? 'Master scheduler' : 'Child scheduler'
+
+            reload_schedule!
+            if am_master
```

Comment:
This if block appears to be a major change. `reload_schedule!` stops the old and creates a new rufus scheduler, but does not appear to start it. It looks like we start here immediately if master, and stop if not. I think this is related to the below changes in this file I have called out. It appears only the master scheduler is running scheduled tasks, but all children are running delayed? Not totally clear on this.

Reply:
That's correct. We can't run rufus-scheduler on more than one process or else they'll enqueue duplicate jobs. So we use the master lock to decide which process gets to run rufus-scheduler. The master does both scheduled and delayed jobs, and all other workers do only delayed jobs.

Note that the load imposed by scheduled jobs is much lighter than the load from delayed jobs, so performance isn't really a concern there.